apps/jsonrpc2/test/jsonrpc2/handler_test.exs

```diff
@@ -0,0 +1,240 @@
+defmodule JSONRPC2.HandlerTest do
+  use ExUnit.Case, async: true
+
+  import ExUnit.CaptureLog
+  alias JSONRPC2.BuggyHandlerTest
+  alias JSONRPC2.ErrorHandlerTest
+  alias JSONRPC2.SpecHandlerTest
+
+  describe "examples from JSON-RPC 2.0 spec at http://www.jsonrpc.org/specification#examples" do
+    test "rpc call with positional parameters" do
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}),
+        ~s({"jsonrpc": "2.0", "result": 19, "id": 1})
+      )
+
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2}),
+        ~s({"jsonrpc": "2.0", "result": -19, "id": 2})
+      )
+    end
+
+    test "rpc call with named parameters" do
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}),
+        ~s({"jsonrpc": "2.0", "result": 19, "id": 3})
+      )
+
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "subtract", "params": {"minuend": 42, "subtrahend": 23}, "id": 4}),
+        ~s({"jsonrpc": "2.0", "result": 19, "id": 4})
+      )
+    end
+
+    test "a Notification" do
+      assert_rpc_noreply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "update", "params": [1,2,3,4,5]})
+      )
+
+      assert_rpc_noreply(SpecHandlerTest, ~s({"jsonrpc": "2.0", "method": "foobar"}))
+    end
+
+    test "rpc call of non-existent method" do
+      assert_rpc_reply(
+        SpecHandlerTest,
+        ~s({"jsonrpc": "2.0", "method": "foobar", "id": "1"}),
+        ~s({"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": {"method": "foobar", "params": []}}, "id": "1"})
```

Comment:
It doesn't matter much, but was thinking more like 
```
  @parser_error_code -32_700
  @invalid_request_code -32_600
  @method_not_found_code -32_601
...
```
The idea is that the test should verify that the error code is received but shouldn't care what the value of the error code is. Since they're not likely to change much, it probably doesn't matter.

Reply:
https://github.com/mana-ethereum/mana/pull/600/files#diff-c81349a8dc0508cbfcd6768895910d7dR37